### PR TITLE
SOV-1309: reward manager integration

### PR DESCRIPTION
**Pull request by dharkmattr**
Resolves https://sovryn.atlassian.net/browse/SOV-1309

### CODE REVIEW

File: src/config/contracts.ts [hunk 1]
```diff
@@ -32,6 +33,7 @@ export const contractsAddresses: ContractsAddresses = {
     vestingRegistry: '0x036ab2DB0a3d1574469a4a7E09887Ed76fB56C41',
     XUSDMassetProxy: '0x1440d19436bEeaF8517896bffB957a88EC95a00F',
     pauseManager: '0x3e96B99993B39a01e620Be71eFB5bF8AC6F7fb0E',
+    rewardManager: '',
```
Comment (tiltom) [inline src/config/contracts.ts:36]: Add a `TODO` comment so you don't forget about it.

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 2]
```diff
@@ -85,6 +87,10 @@ export const AggregatorComponent = ({
     setValue
   );
 
+  useEffect(() => {
+    setValue<AggregatorInputs.ReceiveAmount>(AggregatorInputs.ReceiveAmount, receiveAmount || '0.0');
```
Comment (tiltom) [inline src/pages/Aggregator/Aggregator.component.tsx:91]:
```suggestion
    setValue(AggregatorInputs.ReceiveAmount, receiveAmount || '0.0');
```

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 3]
```diff
@@ -136,7 +142,6 @@ export const AggregatorComponent = ({
 
   useEffect(() => {
     resetField(AggregatorInputs.SendAmount);
-    resetField(AggregatorInputs.ReceiveAmount);
```
Comment (tiltom) [inline src/pages/Aggregator/Aggregator.component.tsx:139]: Why did you remove it? Now you can end up in a state like this.
image

File: src/pages/Aggregator/Aggregator.fields.ts [hunk 4]
```diff
@@ -8,8 +8,8 @@ export enum AggregatorInputs {
   SendAmount = 'sendAmount',
   DestinationChain = 'destinationChain',
   DestinationToken = 'destinationToken',
-  ReceiveAmount = 'receiveAmount',
   ReceiveAddress = 'receiveAddress',
+  ReceiveAmount = 'receiveAmount'
```
Comment (tiltom) [inline src/pages/Aggregator/Aggregator.fields.ts:12]: You can leave it as-is now but this just produces unnecessary changes and makes PRs harder to review since there was no real change in this file.

File: src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx [hunk 5]
```diff
@@ -32,6 +35,22 @@ export const AggregatorInfoComponent = ({
...
+        gridTemplateColumns: 'repeat(2, 1fr)',
+        gap: '10px 8px',
+        mt: 8,
+        whiteSpace: 'nowrap',
+        color: ({ palette }) => palette.grey[600],
+      }}
+    >
+      <InfoRow
+        label={incentives?.type === IncentiveType.penalty ? 'Penalty' : 'Reward'}
+        value={`${incentives?.amount || '0.0'} XUSD`}
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:50]: Mocks in the story contains also a percentage value, discuss it with John.

File: src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx [hunk 6]
```diff
@@ -32,6 +35,22 @@ export const AggregatorInfoComponent = ({
...
+        state={incentivesState || 'idle'}
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:51]:
```suggestion
        state={incentivesState}
```
No need for the fallback since `idle` is initial incentive state.

File: src/pages/Aggregator/AggregatorInfo/AggregatorInfo.types.ts [hunk 7]
```diff
@@ -7,8 +7,11 @@ export type AggregatorInfoContainerProps = {
 
 export type AggregatorInfoComponentProps = {
   onClick: () => void;
-  state: FiniteStates;
+  feesAndLimitsState: FiniteStates;
   feesAndLimits: FeesAndLimitsType;
   tokenName?: string;
   tokenDecimals?: number;
+  incentivesState?: FiniteStates;
+  incentives?: IncentivesType;
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/AggregatorInfo.types.ts:15]: Why are these optional? You always have `incentivesState` (default value is `idle`) and `IncentivesType` is the same as `FeesAndLimitsType` (both contains optional properties).

File: src/store/aggregator/aggregator.sagas.ts [hunk 8]
```diff
@@ -121,6 +128,58 @@ export function* fetchPausedTokens() {
...
+    const rewardManager = yield* select(rewardManagerSelector);
+    const sendAmount = yield* select(sendAmountSelector);
+    const flowState = yield* select(flowStateSelector);
+
+    if (!rewardManager) {
+      throw new Error('Could not find RewardManager contract');
+    }
+
+    const startingTokenAddress = (yield* select(startingTokenAddressSelector)) ?? '';
+    const destinationTokenAddress = (yield* select(destinationTokenAddressSelector)) ?? '';
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:145]:
```suggestion
    const startingTokenAddress = yield* select(startingTokenAddressSelector);
    const destinationTokenAddress = yield* select(destinationTokenAddressSelector);
```
No need for fallbacks, you check them before usage.

File: src/store/aggregator/aggregator.sagas.ts [hunk 9]
```diff
@@ -121,6 +128,58 @@ export function* fetchPausedTokens() {
...
+    let incentiveType = IncentiveType.none;
+
+    if(flowState === 'deposit' && startingTokenAddress) {
+
+      incentiveType = IncentiveType.reward;
+      const tokenDecimals = yield* select(startingTokenDecimalsSelector);
+      const amount = utils.parseUnits(sendAmount, tokenDecimals);
+      incentive = (yield* call(rewardManager.getRewardForDeposit,
+        startingTokenAddress.toLowerCase(), amount)) as BigNumber;
+      incentive = roundBN(incentive, 3);
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:158]: I don't think you should round `incentive`/`penalty` before you `add`/`subtract` them from `receiveAmount` because you'll lose some precision there.

File: src/store/aggregator/aggregator.sagas.ts [hunk 10]
```diff
@@ -121,6 +128,58 @@ export function* fetchPausedTokens() {
...
+    }
+
+    const startingTokenAddress = (yield* select(startingTokenAddressSelector)) ?? '';
+    const destinationTokenAddress = (yield* select(destinationTokenAddressSelector)) ?? '';
+
+    let incentive = BigNumber.from(0);
+    let receiveAmount = BigNumber.from(0);
+    let incentiveType = IncentiveType.none;
+
+    if(flowState === 'deposit' && startingTokenAddress) {
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:151]: I think you can unify these 2 `if`s because they are almost identical. You just need to extract `if` conditions to separate constants and use few ternary operators (for `incentiveType`, `incentiveMethod` and `receiveAmount`).

File: src/store/aggregator/aggregator.sagas.ts [hunk 11]
```diff
@@ -152,6 +211,15 @@ export function* aggregatorSaga() {
...
     takeLatest(aggregatorActions.setStartingToken, fetchStartingTokenBalance),
 
+    takeLatest(aggregatorActions.setSendAmount, fetchIncentive),
+    takeLatest(aggregatorActions.setDestinationToken, fetchIncentive),
+
     takeLatest(aggregatorActions.submit, transferTokens),
   ]);
 }
+
+function roundBN(bn: BigNumber, digits: number): BigNumber {
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:221]: This definitely doesn't belong here. Create a new file `mathUtils.ts` (if you don't find better existing files) and move it there.

File: src/store/aggregator/aggregator.state.ts [hunk 12]
```diff
@@ -60,6 +78,7 @@ export class AggregatorState {
     state: 'idle',
     data: undefined,
   };
+  sendAmount: string = '';
```
Comment (tiltom) [inline src/store/aggregator/aggregator.state.ts:81]: Make it optional with default value `undefined`, the same as `receiveAmount`.

**Review by tiltom — COMMENTED** [review]
I have 2 more questions:
1. Why did you rename tests? If you did it just because they're failing, rename them back.
2. I see in the story there should be a slippage slider, have you decided with John you don't need it or why it's not implemented?

File: src/store/aggregator/aggregator.sagas.ts (hunk 10 above)
Comment (dharkmattr) [reply src/store/aggregator/aggregator.sagas.ts:151]: tried it, and I think it's more confusing....

**Comment by tiltom** [comment]
I did functional testing and I'll do a code review once the following things are fixed:

1. If starting/destination chain is other than `RSK`, there should be no slippage tolerance or reward amount displayed.
Pasted Graphic 4
Pasted Graphic 6

2. I always see “Reward” with the value of `0` if I select `XUSD` as a starting asset. Either it’s not working correctly or we need to set some withdrawal penalties on testnet to test it.

3. The same goes for a `bAsset` -> `XUSD` direction. I also don’t understand why we still have disabled deposits for some assets if the balancing curves should replace that mechanism.
Pasted Graphic 7

4. The PR broke converting functionality - we no longer see receive amount when we enter the amount we want to convert. That causes the “Convert” button to be disabled and we cannot do anything.
Pasted Graphic 9

@dharkmattr let me know if anything is unclear/you want to consult something.

File: package.json [hunk 13]
```diff
@@ -42,16 +43,19 @@
...
     "redux-persist": "6.0.0",
     "redux-saga": "1.1.3",
     "rxjs": "7.8.1",
     "subscription-client": "0.9.15",
     "ts-node": "10.5.0",
-    "typed-redux-saga": "1.4.0"
+    "typed-redux-saga": "^1.5.0",
+    "typescript-http-client": "^0.10.4",
+    "web3": "^1.9.0",
+    "web3-utils": "^1.10.0"
```
Comment (soulBit) [inline package.json:58]:
```suggestion
    "web3-utils": "1.10.0"
```
Please remove carets from package versions, we would like to use static versions if possible

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 14]
```diff
@@ -85,6 +97,12 @@ export const AggregatorComponent = ({
     setValue
   );
 
+  const slippageProtectionAvailable = startingChain === destinationChain;
+
+  useEffect(() => {
+    setValue(AggregatorInputs.ReceiveAmount, receiveAmount || '0.0');
```
Comment (soulBit) [inline src/pages/Aggregator/Aggregator.component.tsx:103]: When a low amount of funds are converted, the penalty amount is not visible but I can see it applied to receive amount:
image

**Review by soulBit — CHANGES_REQUESTED** [review]

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 15]
```diff
@@ -40,12 +44,21 @@ const PageViewTitle: React.FC = ({ children }) => (
   </Box>
 );
 
+const slippageSliderValues = [0, 0.5, 1, 1.5, 2, 2.5];
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/pages/Aggregator/Aggregator.component.tsx:47]: Gotta update the spec

File: src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx [hunk 16]
```diff
@@ -36,7 +41,32 @@ export const AggregatorInfoComponent = ({
...
+          !(
+            incentives?.type === IncentiveType.penalty ||
+            incentives?.type === IncentiveType.reward
+          )
+        }
+        label={
+          incentives?.type === IncentiveType.penalty ? 'Penalty' : 'Reward'
+        }
+        value={`${incentives?.amount || '0.0'}`}
+        total={`${sendAmount || '0.0'}`}
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:65]: What do you mean "inconsistent with behaviour for RSK -> RSK conversions"?

[the remark this answers is not in the document]
Comment (soulBit) [reply src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:65]: The behaviour in 1st screenshot (Eth -> RSK) is different to the 2nd (RSK -> RSK), when an asset is locked - I believe it should work the same in both cases

[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:65]: The difference is between the two tokens. One has a reward of 0 and one a reward of 0.006. IMHO that's an actual difference

[the remark this answers is not in the document]
Comment (soulBit) [reply src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:65]: Ah that makes sense then, marking as resolved

File: src/store/aggregator/sagas/depositTokens.ts [hunk 17]
```diff
@@ -88,12 +92,19 @@ export function* depositTokens({ payload }: AggregatorActions['submit']) {
...
+  if (!isCrossChain) {
+    let minimumRewardNumber = rewardAmount * (100 - payload.slippageSlider) / 100;
+    minimumRewardNumber = minimumRewardNumber > 0 ? minimumRewardNumber : 0;
+    const minimumReward = utils.parseUnits(
+      minimumRewardNumber.toString(),
+      DEFAULT_ASSET_DECIMALS
+    );
     submitEffect = call(
-      massetContract.mintTo,
+      massetContract.mintToWithMinimumReward,
```
Comment (soulBit) [inline src/store/aggregator/sagas/depositTokens.ts:103]: RUSDT -> XUSD conversions are failing with "reward under minimum" error:
image

https://dashboard.tenderly.co/tx/rsk-testnet/0x777438c456bc7da900cf188ade77392b9d685c07e5dde727a98313306a6c272d

this was working previously, but I just noticed that in my successful test prior to this that I never actually received the reward amount I was shown (should have got ~1.0006 XUSD):
https://explorer.testnet.rsk.co/tx/0xf7a0313164a4ba39664d897b73811e3a8c630b18b2fcd46601da35ee411f9913?__ctab=Token%20Transfers

File: src/store/aggregator/sagas/withdrawTokens.ts [hunk 18]
```diff
@@ -77,16 +81,22 @@ export function* withdrawTokens({ payload }: AggregatorActions['submit']) {
 
   let submitEffect: SagaContractEffect;
 
-  if (isRSK) {
+  if (!isCrossChain) {
+    const maximumPenaltyNumber = penaltyAmount * (100 + payload.slippageSlider) / 100;
```
[the remark this answers is not in the document]
Comment (soulBit) [reply src/store/aggregator/sagas/withdrawTokens.ts:85]: @dharkmattr conversion now fails to initiate with an underflow error shown in console, please take a look (starting network here is RSK testnet also):
image

File: src/store/aggregator/sagas/withdrawTokens.test.ts [hunk 19]
```diff
@@ -97,12 +97,7 @@ describe('withdrawTokens', () => {
         .provide([
           ...mockSelectors,
           [matchers.call.fn(mockToken.allowance), parseUnits('11')],
-          [
-            matchers.call.fn(
-              mockMasset['redeemToBridge(address,uint256,address)']
-            ),
-            mockTx,
-          ],
+          [matchers.call.fn(mockMasset.redeemToBridge), mockTx],
```
[the remark this answers is not in the document]
Comment (soulBit) [reply src/store/aggregator/sagas/withdrawTokens.test.ts:100]: this error is now resolved, however I see underflow error when trying to initiate conversion as per my other comment

**Review by soulBit — CHANGES_REQUESTED** [review]

File: src/pages/Aggregator/Aggregator.component.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (john-light) [reply src/pages/Aggregator/Aggregator.component.tsx:47]: @dharkmattr why should the spec be updated?

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 20]
```diff
@@ -218,6 +255,17 @@ export const AggregatorComponent = ({
...
                 setValue={setValue}
               />
+              <ControlledSlider
+                hidden={!slippageProtectionAvailable}
+                title="Slippage Tolerance"
+                name={AggregatorInputs.SlippageSlider}
+                control={control}
+                setValue={setValue}
+                min={0}
+                max={2.5}
```
[the remark this answers is not in the document]
Comment (tiltom) [reply src/pages/Aggregator/Aggregator.component.tsx:265]: This is not resolved as well, discuss with @john-light.

File: src/pages/Aggregator/Aggregator.fields.ts [hunk 21]
```diff
@@ -28,6 +30,7 @@ export const aggregatorDefaultValues: DefaultValues<AggregatorFormValues> = {
   [AggregatorInputs.SendAmount]: '',
   [AggregatorInputs.DestinationChain]: '',
   [AggregatorInputs.DestinationToken]: '',
-  [AggregatorInputs.ReceiveAmount]: '',
   [AggregatorInputs.ReceiveAddress]: '',
+  [AggregatorInputs.ReceiveAmount]: '',
+  [AggregatorInputs.SlippageSlider]: 2.5,
```
[the remark this answers is not in the document]
Comment (tiltom) [reply src/pages/Aggregator/Aggregator.fields.ts:35]: Again - discuss with @john-light.

File: src/pages/Aggregator/Aggregator.utils.ts [hunk 22]
```diff
@@ -33,3 +43,67 @@ export const removeFocusFromCurrentField = ({
...
+}): boolean {
+  const startingTokenBalance = useSelector(startingTokenBalanceSelector) ?? 0;
+  const destinationTokenAggregatorBalance = useSelector(
+    destinationTokenAggregatorBalanceSelector
+  ) ?? 0;
+  const startingTokenDecimals = useSelector(startingTokenDecimalsSelector);
+
+  // validate the tokens
+  if (
+    !(isRSK(state.startingChain) || isRSK(state.destinationChain)) ||
```
Comment (tiltom) [inline src/pages/Aggregator/Aggregator.utils.ts:65]: This is definitely not correct - you would basically validate every conversion except `RSK -> RSK` as invalid.

File: src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx [hunk 23]
```diff
@@ -36,7 +41,32 @@ export const AggregatorInfoComponent = ({
...
+      sx={{
+        display: 'grid',
+        gridTemplateColumns: 'repeat(2, 1fr)',
+        gap: '10px 8px',
+        mt: 8,
+        whiteSpace: 'nowrap',
+        color: ({ palette }) => palette.grey[600],
+      }}
+    >
+      <InfoRowWithPerc
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/AggregatorInfo.component.tsx:54]:
```suggestion
      <InfoRowWithPercentageValue
```

File: src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx [hunk 24]
```diff
@@ -0,0 +1,64 @@
+import Skeleton from '@mui/material/Skeleton';
+import Typography from '@mui/material/Typography';
+import { FiniteStates } from '../../../utils/types';
+
+type InfoRowWithPercProps = {
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx:5]:
```suggestion
type InfoRowWithPercentageValueProps = {
```

File: src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+  hidden?: boolean;
+  label: string;
+  value?: string;
+  total?: string;
+  unit?: String;
+  state: FiniteStates;
+};
+
+export const InfoRowWithPerc = ({
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx:14]:
```suggestion
export const InfoRowWithPercentageValue = ({
```

File: src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+  label,
+  value,
+  unit,
+  total,
+  state,
+  hidden,
+}: InfoRowWithPercProps) => {
+  const isUpdate = !value || state === 'loading';
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx:22]: Use `useMemo`.

File: src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+          }}
+        >
+          {value} {unit} {calculatePercStr(value, total)}
+        </Typography>
+      )}
+    </>
+  );
+};
+
+function calculatePercStr(value?: string, total?: string): string {
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx:58]:
```suggestion
function calculatePercentageStr(value?: string, total?: string): string {
```

File: src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+  if (Number(total) === 0) return '';
+  const tv = Number(value ?? '');
+  const tt = Number(total ?? '');
+  const p = Math.round((10000 * tv) / tt) / 100;
+  return p > 0 ? `(${p}%)` : '';
```
Comment (tiltom) [inline src/pages/Aggregator/AggregatorInfo/InfoRowWithPerc.component.tsx:63]: Don't use abbreviations, rename variables to something more readable.

File: src/pages/Aggregator/SlippageSlider/SlippageSlider.component.tsx [hunk 29]
```diff
@@ -0,0 +1,34 @@
+import Box from '@mui/material/Box';
+import Typography from '@mui/material/Typography';
+import Slider from '@mui/material/Slider';
+
+import { SlippageSliderProps } from './SlippageSlider.types';
+
+export const SlippageSlider = ({
```
Comment (tiltom) [inline src/pages/Aggregator/SlippageSlider/SlippageSlider.component.tsx:7]:
```suggestion
export const Slider = ({
```
The component is general enough so it can be used for any kind of slider, not just for slippage. Rename the rest of the files (`.controlled`, `.types`) as well.

File: src/store/aggregator/aggregator.sagas.ts [hunk 30]
```diff
@@ -169,7 +182,84 @@ export function* fetchPausedTokens() {
...
+      incentive = (yield* call(
+        getPenalty,
+        sovTokenAddress!,
+        amount,
+        isMainnet
+      )) as BigNumber;
+      receiveAmount = amount.sub(incentive);
+    }
+
+    //incentive = incentive && roundBN(incentive, 3);
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:248]:
```suggestion

```

File: src/store/aggregator/aggregator.sagas.ts [hunk 31]
```diff
@@ -169,7 +182,84 @@ export function* fetchPausedTokens() {
...
+
+    yield* put(
+      aggregatorActions.setIncentives({
+        type: incentiveType,
+        amount: utils.formatUnits(incentive, DEFAULT_ASSET_DECIMALS),
+      })
+    );
+    yield* put(
+      aggregatorActions.setReceiveAmount(
```
Comment (tiltom) [inline src/store/aggregator/aggregator.sagas.ts:257]: I think that this leads to bugs - you need to set receive amount also in case there is no reward/penalty because only `RSK -> RSK` conversions will work.

File: src/store/aggregator/aggregator.slice.ts [hunk 32]
```diff
@@ -121,6 +152,10 @@ const aggregatorSlice = createSlice({
       state.startingTokenBalance.state = 'idle';
       state.startingTokenBalance.data = undefined;
       state.txDetails = undefined;
+      state.incentives.state = 'idle';
+      state.incentives.data = {};
+      state.receiveAmount = undefined;
+      state.slippageSlider = 2.5;
```
Comment (tiltom) [inline src/store/aggregator/aggregator.slice.ts:158]: Does not match spec, discuss with @john-light.

File: src/store/aggregator/getIncentive.ts [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+        },
+        'latest',
+      ],
+    },
+  });
+  const responseBody = await client.execute<JsonRpcResponse>(request);
+  return responseBody;
+}
+
+export async function getReward(
```
Comment (tiltom) [inline src/store/aggregator/getIncentive.ts:52]: This hack is probably not needed anymore, fix all my other comments first and then try to make it working as I described in this comment: https://sovryn.atlassian.net/browse/SOV-1309?focusedCommentId=17176.

File: src/store/aggregator/sagas/withdrawTokens.ts [hunk 34]
```diff
@@ -29,6 +32,8 @@ export function* withdrawTokens({ payload }: AggregatorActions['submit']) {
   const massetContract = yield* select(massetContractSelector);
   const account = yield* select(accountSelector);
   const { destinationChain, receiveAddress, sendAmount } = payload;
+  const penaltyData = yield* select(incentivesSelector);
+  const penaltyAmount = Number(penaltyData?.amount ?? 0);
```
Comment (tiltom) [inline src/store/aggregator/sagas/withdrawTokens.ts:36]:
```suggestion

```

File: src/store/aggregator/sagas/withdrawTokens.ts [hunk 35]
```diff
@@ -1,17 +1,20 @@
-import { utils } from 'ethers';
+import { BigNumber, utils } from 'ethers';
```
Comment (tiltom) [inline src/store/aggregator/sagas/withdrawTokens.ts:1]:
```suggestion
import { utils } from 'ethers';
```

**Review by tiltom — CHANGES_REQUESTED** [review]

**Comment by tiltom** [comment]
@dharkmattr I found another issue - I cannot convert between RSK and other chains. Once I enter send amount, receive amount is not automatically populated and after clicking “Convert”, I just receive a validation error. As I said in some of the comments below, I believe it's caused by incorrect validation and the way how the receive amount is set.

Pasted Graphic

File: src/pages/Aggregator/Aggregator.component.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/pages/Aggregator/Aggregator.component.tsx:47]: The meaning of these numbers is the point change in fee. So if, for example, I deposit $1000 and get a reward of $1, the maximum tolerance of 1% means that a transaction in which the reward is lesser than $0.99 will fail. That's a slim tolerance and would mean, most likely, that most transactions will fail. 

Thinking from the user's perspective, why would I not be allowed to make a deposit and just be happy with whatever I get?

File: src/config/contracts.ts [hunk 36]
```diff
@@ -30,6 +31,8 @@ export const contractsAddresses: ContractsAddresses = {
     feeSharingProxy: '0x12B1B0C67d9A771EB5Db7726d23fdc6848fd93ef',
     vestingRegistry: '0x036ab2DB0a3d1574469a4a7E09887Ed76fB56C41',
     XUSDMassetProxy: '0x1440d19436bEeaF8517896bffB957a88EC95a00F',
+    // TODO: update with address once deployed to mainnet
+    rewardManager: '',
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/config/contracts.ts:35]: I don't know. If we do have it on mainnet, that should be the subject of a separate ticket and a separate PR.

File: src/pages/Aggregator/Aggregator.component.tsx [hunk 37]
```diff
@@ -85,6 +97,12 @@ export const AggregatorComponent = ({
     setValue
   );
 
+  const slippageProtectionAvailable = startingChain === destinationChain;
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/pages/Aggregator/Aggregator.component.tsx:100]: I'm not sure I understand what's weird about it....

File: src/store/aggregator/getIncentive.ts (hunk 33 above)
Comment (dharkmattr) [reply src/store/aggregator/getIncentive.ts:52]: We are no longer in a position to spend time changing things to no functional advantage. This works. If you think that the other way is "less hacky", we should do that as a separate ticket.

File: src/store/aggregator/sagas/withdrawTokens.ts (hunk 34 above)
Comment (dharkmattr) [reply src/store/aggregator/sagas/withdrawTokens.ts:36]: I don't understand this comment.

File: src/store/aggregator/aggregator.sagas.ts (hunk 31 above)
Comment (dharkmattr) [reply src/store/aggregator/aggregator.sagas.ts:258]: I don't understand why you think so. Non rsk-rsk transactions also have reward/penalty.

File: src/store/aggregator/getIncentive.ts (hunk 33 above)
Comment (tiltom) [reply src/store/aggregator/getIncentive.ts:52]: Why do you say we're in a rush now? @john-light can you confirm we should go with the current approach even though it means we'll introduce some technical debt? I think we should do our best and not make the codebase worse than already is.

File: src/store/aggregator/sagas/withdrawTokens.ts (hunk 34 above)
Comment (tiltom) [reply src/store/aggregator/sagas/withdrawTokens.ts:36]: If you open it in IDE you will. `penaltyAmount` is not used anywhere.

File: src/config/contracts.ts (hunk 36 above)
[the remark this answers is not in the document]
Comment (tiltom) [reply src/config/contracts.ts:35]: I don't think it should be a separate ticket/PR because you were claiming we're ready for production any day now but how can we be ready for that if we haven't do a single test of it in production?

File: src/pages/Aggregator/Aggregator.component.tsx (hunk 37 above)
[the remark this answers is not in the document]
Comment (tiltom) [reply src/pages/Aggregator/Aggregator.component.tsx:100]: You shouldn't mix hook variables and non-hook variables as hooks should be called at the top level of a component. Also, you usually want to use `useMemo`. This means you have 3 options:

1.  Use `useMemo` and leave the variable in the same place.
2. Move the variable below the last hook variable.
3. Use it inline.

File: src/store/aggregator/getIncentive.ts (hunk 33 above)
Comment (dharkmattr) [reply src/store/aggregator/getIncentive.ts:52]: @john-light is our product manager, not our technical decision maker. As this is a technical decision, I think I'll make the call. I don't disagree with you, I just think that this is an improvement that can and should be put in a ticket and done as a separate PR. The reasons are: 1. time is short, 2. the current solution is localized, clear, and tested, and 3. there is technical debt in this project that should be fixed and is much higher priority - specifically the failing tests, low test coverage, and convoluted logic.

File: src/config/contracts.ts (hunk 36 above)
[the remark this answers is not in the document]
Comment (dharkmattr) [reply src/config/contracts.ts:35]: You're absolutely right, I was looking at the wrong thing.

File: src/pages/Aggregator/Aggregator.component.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (john-light) [reply src/pages/Aggregator/Aggregator.component.tsx:47]: @dharkmattr the values were copied from the slippage tolerance settings used for the Sovryn AMM. They were a first approximation of what the numbers should be, prior to knowing what the actual reward/penalty parameters will be. After considering the discussions we've had about this since the PRD was first written last year, and considering your suggestion above, I propose this adjustment to the spec for this UI feature:

- The slippage tolerance setting should define the maximum difference from the current reward/penalty applied by the smart contract e.g. if the user is depositing 1000 USDT and would receive a reward of 25 XUSD then a slippage tolerance setting of 5% would mean that the minimum reward the user will accept is 23.75 XUSD, otherwise the tx will revert.
  - This setting should only appear for transactions with Rootstock as the starting chain.
- The slippage tolerance section should include an amount entry field like the token amount entry field above it, with default selector buttons below with the values `1%, 1.5%, 2%, 2.5%, MAX`.
  - The user can either manually enter a value into the entry field, or click one of the selector buttons to enter a default value. If no value is entered, then the `MAX` value should be passed with the transaction.
  - `MAX` should result in a reward or penalty % equal to the `Max reward` or `Max penalty` based on the parameter in the contract, as applicable for the given chain/token selected.
- The user should be shown the `Penalty` and `Maximum penalty`, or `Reward` and `Minimum reward`, as applicable based on their selected slippage tolerance setting. This value should include the % of the reward or penalty which is equal to `Reward or penalty amount / Conversion amount` e.g. based on the example above, the user would see a `Estimated reward` of `25 XUSD (2.5%)` and `Minimum reward` of `23.75 XUSD (2.375%)`

WDYT?

Mockup:
BF Slippage

[the remark this answers is not in the document]
Comment (john-light) [reply src/pages/Aggregator/Aggregator.component.tsx]: We decided to remove the slippage tolerance setting for now and use a sane default that is applied in the background. Spec on Jira has been updated to reflect this. We may re-implement the slippage tolerance setting as specified above in a future update.

File: src/pages/Aggregator/Aggregator.component.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (john-light) [reply src/pages/Aggregator/Aggregator.component.tsx]: We decided to remove the slippage tolerance setting for now and use a sane default that is applied in the background. Spec on Jira has been updated to reflect this.

File: src/pages/Aggregator/Aggregator.fields.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (john-light) [reply src/pages/Aggregator/Aggregator.fields.ts]: We decided to remove the slippage tolerance setting for now and use a sane default that is applied in the background. Spec on Jira has been updated to reflect this.

File: src/store/aggregator/aggregator.slice.ts (hunk 32 above)
Comment (john-light) [reply src/store/aggregator/aggregator.slice.ts]: We decided to remove the slippage tolerance setting for now and use a sane default that is applied in the background. Spec on Jira has been updated to reflect this.

File: src/store/aggregator/getIncentive.ts (hunk 33 above)
Comment (john-light) [reply src/store/aggregator/getIncentive.ts:52]: > @john-light is our product manager, not our technical decision maker. As this is a technical decision, I think I'll make the call.

While I appreciate the consideration to ask for my opinion on how this should be implemented, as long as the requirements in the PRD/product spec are satisfied it is @dharkmattr's role to make the final call on _how_ to technically satisfy the requirements. So on this I will defer to his decision.
